### PR TITLE
CSHARP-4272: Validate return value of bsonClassMap.CreateInstance() in BsonClassMapSerializer

### PR DESCRIPTION
**Pull request by BorisDog**

### CODE REVIEW

File: src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs [hunk 1]
```diff
@@ -139,6 +139,11 @@ public TClass DeserializeClass(BsonDeserializationContext context)
                 // for mutable classes we deserialize the values directly into the result object
                 document = (TClass)_classMap.CreateInstance();
 
+                if (document == null)
+                {
+                    throw new InvalidOperationException($"{nameof(BsonClassMap)} did not provide an instance of {typeof(TClass).Name}.");
```
Comment (JamesKovacs) [inline src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs:144]: Reviewing the exceptions thrown in `BsonClassMapSerializer`, I think that `BsonSerializationException` is more appropriate than `InvalidOperationException`. `BsonSerializationException` is used when a creator map is unexpectedly null or similar. `InvalidOperationException` is thrown when we attempt to perform an invalid operation such as setting an Id when there is no setter.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Minor change with regard to the type of exception to throw.

File: src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs (hunk 1 above)
Comment (BorisDog) [reply src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs]: Done.

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**